audit_log/README.md

```diff
@@ -0,0 +1,207 @@
+- **Extensible:** Provides a base class for creating custom audit log services tailored to specific needs.
+- **Admin Integration:** Seamlessly integrates with Django admin to audit model operations.
+- **Optimized for Performance:** Includes a specialized paginator for handling large audit log tables.
+
+### Audit Log Model
+
+The `AuditLogEntry` [model](./models.py) stores audit log entries in the database. Each entry represents a single audit event and contains the following fields:
+
+Fields:
+
+- `is_sent`: Indicates whether the audit log entry has been sent to an external system or processed in any way.
+- `message`: Stores the audit log message in a JSON format (see the ["Audit Log Message Format"](#audit-log-message-format) section for details). This field contains the core information about the audit event, including the actor, action, target, and timestamp.
+- `created_at`: Records the timestamp when the audit log entry was created.
+
+This model provides a structured way to store and retrieve audit log data, allowing for efficient querying and analysis.
+
+### Audit Log Message Format
+
+Audit log messages are stored in a JSON format with the following structure:
+
+```json
+{
+  "audit_event": {
+    "actor": {
+      "ip_address": "127.0.0.1",
+      "role": "ADMIN",
+      "uuid": "7a52666c-9b5b-11ef-91f6-e2cd5b1fb5ac"
+    },
+    "date_time": "2024-11-13T14:13:57.853Z",
+    "date_time_epoch": 1731507237853,
+    "operation": "READ",
+    "origin": "notification_service",
+    "status": "SUCCESS",
+    "target": {
+      "object_ids": [
+        "bd6a5d06-8828-47a7-bb5b-fdf4559da56e",
+        "9fb9cfa1-23f4-4401-b66a-8cc9a3277175",
+        "ee444103-f145-4282-8768-28cdee52c3a4",
+        "f928162d-6d27-4fab-a084-fd83a3dd1c5a",
+        "c1bf4a0d-0ae3-4191-8948-94bec1742b41",
+        "33935ff8-e8db-43e6-8939-15974dfd6d74",
+        "e77c9e36-1f24-49dd-a700-2191caf594a4",
+        "77e74bab-91ca-48b3-8245-cfd08196af80"
+      ],
+      "path": "/admin/api/deliverylog/"
+    }
+  }
+}
+```
+
+Fields:
+
+- `audit_event`: Contains the details of the audit event.
+- `actor`: Information about the user who performed the action.
+- `ip_address`: The IP address of the actor.
+- `role`: The role of the actor (e.g., "ADMIN", "USER").
+- `uuid`: The unique identifier of the actor.
+- `date_time`: The date and time of the event in ISO 8601 format.
+- `date_time_epoch`: The date and time of the event in Unix epoch milliseconds.
+- `operation`: The type of operation performed (e.g., "READ", "CREATE", "UPDATE", "DELETE").
+- `origin`: The origin of the event (e.g., "notification_service").
+- `status`: The status of the operation (e.g., "SUCCESS", "FAILURE").
+- `target`: Information about the object(s) affected by the operation.
+- `object_ids`: A list of unique identifiers of the affected objects.
+- `path`: The request path associated with the operation.
```

Comment:
How about adding the nesting levels into this to make the structure clearer (in addition to having an example above this), how do you see this?
```suggestion
- `audit_event`: Contains the details of the audit event.
  - `actor`: Information about the user who performed the action.
    - `ip_address`: The IP address of the actor.
    - `role`: The role of the actor (e.g., "ADMIN", "USER").
    - `uuid`: The unique identifier of the actor.
  - `date_time`: The date and time of the event in ISO 8601 format.
  - `date_time_epoch`: The date and time of the event in Unix epoch milliseconds.
  - `operation`: The type of operation performed (e.g., "READ", "CREATE", "UPDATE", "DELETE").
  - `origin`: The origin of the event (e.g., "notification_service").
  - `status`: The status of the operation (e.g., "SUCCESS", "FAILURE").
  - `target`: Information about the object(s) affected by the operation.
    - `object_ids`: A list of unique identifiers of the affected objects.
  - `path`: The request path associated with the operation.
```